internal/broker/broker_test.go

```diff
@@ -827,6 +830,166 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+				err = password.HashAndStorePassword("password", b.PasswordFilepathForSession(sessionID))
+				require.NoError(t, err, "Setup: HashAndStorePassword should not have returned an error")
+
+				updateAuthModes(t, b, sessionID, authmodes.Password)
+
+				authData := `{"challenge":"` + encryptChallenge(t, "password", key) + `"}`
+
+				access, data, err := b.IsAuthenticated(sessionID, authData)
+				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
+				require.NoError(t, err)
+				require.NotEqual(t, access, broker.AuthDenied, "authentication failed")
+				got := isAuthenticatedResponse{Access: access, Data: data, Err: fmt.Sprint(err)}
+				out, err := yaml.Marshal(got)
+				require.NoError(t, err, "Failed to marshal first response")
+
+				err = os.WriteFile(filepath.Join(outDir, u), out, 0600)
+				require.NoError(t, err, "Failed to write response")
+
+				// Ensure that the token content is generic to avoid golden file conflicts
+				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
+					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic token file")
+				}
+				passwordPath := b.PasswordFilepathForSession(sessionID)
+				if _, err := os.Stat(passwordPath); err == nil {
+					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic password file")
+				}
+			}
+			for _, u := range tc.unallowedUsernames {
+				sessionID, key := newSessionForTests(t, b, u, "")
+				session = sessionID
+				token := tokenOptions{username: u}
+				generateAndStoreCachedInfo(t, token, b.TokenPathForSession(sessionID))
+				err = password.HashAndStorePassword("password", b.PasswordFilepathForSession(sessionID))
+				require.NoError(t, err, "Setup: HashAndStorePassword should not have returned an error")
+
+				updateAuthModes(t, b, sessionID, authmodes.Password)
+
+				authData := `{"challenge":"` + encryptChallenge(t, "password", key) + `"}`
+
+				access, data, err := b.IsAuthenticated(sessionID, authData)
+				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
+				require.NoError(t, err)
+				require.Equal(t, access, broker.AuthDenied, "authentication succeeded")
+
+				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
+					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic token file")
+				}
+				passwordPath := b.PasswordFilepathForSession(sessionID)
+				if _, err := os.Stat(passwordPath); err == nil {
+					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic password file")
+				}
+			}
```

Comment:
We can avoid some code duplication here by iterating over the allowed and unallowed users in the same loop:

```diff
diff --git a/internal/broker/broker_test.go b/internal/broker/broker_test.go
index 55de0d0..4c9d017 100644
--- a/internal/broker/broker_test.go
+++ b/internal/broker/broker_test.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"slices"
 	"testing"
 	"time"
 
@@ -891,13 +892,16 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 			err := os.Mkdir(dataDir, 0700)
 			require.NoError(t, err, "Setup: Mkdir should not have returned an error")
 			var session string
+			var users []string
 
 			idTokenClaims := []map[string]interface{}{}
 			for _, uname := range tc.allowedUsernames {
 				idTokenClaims = append(idTokenClaims, map[string]interface{}{"sub": "user", "name": "user", "email": uname})
+				users = append(users, uname)
 			}
 			for _, uname := range tc.unallowedUsernames {
 				idTokenClaims = append(idTokenClaims, map[string]interface{}{"sub": "user", "name": "user", "email": uname})
+				users = append(users, uname)
 			}
 
 			b := newBrokerForTests(t, &brokerForTestConfig{
@@ -912,7 +916,7 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 				},
 			})
 
-			for _, u := range tc.allowedUsernames {
+			for _, u := range users {
 				sessionID, key := newSessionForTests(t, b, u, "")
 				session = sessionID
 				token := tokenOptions{username: u}
@@ -927,7 +931,13 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 				access, data, err := b.IsAuthenticated(sessionID, authData)
 				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
 				require.NoError(t, err)
-				require.NotEqual(t, access, broker.AuthDenied, "authentication failed")
+
+				if slices.Contains(tc.allowedUsernames, u) {
+					require.Equal(t, access, broker.AuthGranted, "authentication failed")
+				} else {
+					require.Equal(t, access, broker.AuthDenied, "authentication failed")
+				}
+
 				got := isAuthenticatedResponse{Access: access, Data: data, Err: fmt.Sprint(err)}
 				out, err := yaml.Marshal(got)
 				require.NoError(t, err, "Failed to marshal first response")
@@ -946,33 +956,6 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 					require.NoError(t, err, "Teardown: Failed to write generic password file")
 				}
 			}
-			for _, u := range tc.unallowedUsernames {
-				sessionID, key := newSessionForTests(t, b, u, "")
-				session = sessionID
-				token := tokenOptions{username: u}
-				generateAndStoreCachedInfo(t, token, b.TokenPathForSession(sessionID))
-				err = password.HashAndStorePassword("password", b.PasswordFilepathForSession(sessionID))
-				require.NoError(t, err, "Setup: HashAndStorePassword should not have returned an error")
-
-				updateAuthModes(t, b, sessionID, authmodes.Password)
-
-				authData := `{"challenge":"` + encryptChallenge(t, "password", key) + `"}`
-
-				access, data, err := b.IsAuthenticated(sessionID, authData)
-				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
-				require.NoError(t, err)
-				require.Equal(t, access, broker.AuthDenied, "authentication succeeded")
-
-				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
-					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
-					require.NoError(t, err, "Teardown: Failed to write generic token file")
-				}
-				passwordPath := b.PasswordFilepathForSession(sessionID)
-				if _, err := os.Stat(passwordPath); err == nil {
-					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
-					require.NoError(t, err, "Teardown: Failed to write generic password file")
-				}
-			}
 
 			if session != "" {
 				issuerDataDir := filepath.Dir(b.UserDataDirForSession(session))
```